### PR TITLE
Add support for HDC2010 sensor

### PR DESCRIPTION
**Pull request by vigsterkr**
### Contribution description
Add driver for HDC2010 temperature and humidity sensor from TI via I2C

### Testing procedure
Connected the HDC2010 sensor to a samr21-xpro board via I2C and read the measurements via SAUL.

### CODE REVIEW

File: dist/tools/pyterm/pyterm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/local/bin/python3.8
```
Comment (benpicco) [inline dist/tools/pyterm/pyterm:1]: Unrelated change.

File: drivers/hdc2010/Makefile.dep [hunk 2]
```diff
@@ -0,0 +1,6 @@
+USEMODULE += xtimer
+FEATURES_REQUIRED += periph_i2c
+
+ifneq (,$(filter hdc2010_irq,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_gpio_irq
+endif
```
Comment (benpicco) [inline drivers/hdc2010/Makefile.dep:6]: needs a newline at the end

File: drivers/hdc2010/hdc2010.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    i2c_release(dev->p.i2c);
+    return status;
+}
+
+int hdc2010_get_results(const hdc2010_t *dev, int16_t *temp, int16_t *hum)
+{
+    int status = HDC2010_OK;
+    assert(dev);
+
```
Comment (benpicco) [inline drivers/hdc2010/hdc2010.c:123]: Shouldn't the

```C
    xtimer_mutex_lock_timeout(&dev->mutex, CONFIG_HDC2010_CONVERSION_TIME);
```

be in here?
And is there really a benefit in making `hdc2010_trigger_conversion()`/`hdc2010_get_results()` public?

File: drivers/include/hdc2010.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+ * @brief   Parameters needed for device initialization
+ */
+typedef struct {
+    i2c_t i2c;                  /**< bus the device is connected to */
+    uint8_t addr;               /**< address on that bus */
+    gpio_t irq_pin;             /**< IRQ pin to check data ready */
+    hdc2010_res_t res;          /**< resolution used for sampling temp and hum */
+    hdc2010_meas_mode_t mode;   /**< sensors to be used */
+    hdc2010_amm_t amm;          /**< Auto Measurement Mode */
+    hdc2010_int_cb_t user_cb;   /**< User defined interrupt callback */
```
Comment (benpicco) [inline drivers/include/hdc2010.h:139]: What would that be used for?
Can't think of a reason why a user would hook the data ready interrupt of the sensor.

File: drivers/include/hdc2010.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ * finished and the get the results from the device.
+ *
+ * @param[in]  dev          device descriptor of sensor
+ * @param[out] temp         temperature [in 100 * degree centigrade]
+ * @param[out] hum          humidity [in 100 * percent relative]
+ *
+ * @return                  HDC2010_OK on success
+ * @return                  HDC2010_BUSERR on I2C communication failures
+ */
+int hdc2010_read(hdc2010_t *dev, int16_t *temp, int16_t *hum);
```
Comment (benpicco) [inline drivers/include/hdc2010.h:201]: hum can't be negative, right?

File: drivers/include/hdc2010.h [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+ * @brief   Read conversion results for temperature and humidity
+ *
+ * @param[in]  dev          device descriptor of sensor
+ * @param[out] temp         temperature [in 100 * degree centigrade]
+ * @param[out] hum          humidity [in 100 * percent relative]
+ *
+ * @return                  HDC2010_OK on success
+ * @return                  HDC2010_BUSERR on I2C communication failures
+ */
+int hdc2010_get_results(const hdc2010_t *dev, int16_t *temp, int16_t *hum);
```
Comment (benpicco) [inline drivers/include/hdc2010.h:186]: I'd just drop those from the public interface to keep things simple.

File: dist/tools/pyterm/pyterm (hunk 1 above)
Comment (vigsterkr) [reply dist/tools/pyterm/pyterm:1]: sorry this got into from my dev env...

File: drivers/include/hdc2010.h (hunk 4 above)
Comment (vigsterkr) [reply drivers/include/hdc2010.h:139]: so actually the story here is the following - maybe i should rather use this usecase in the test?

when you are using the AMM mode (meaning you specify a predefined frequency of measurement of the values), you will get IRQ from the chip itself when the data is read; so the pseudo code would look something like this:

 1) init the dev with your own callback function, where in the callback function you do a `hdc2010_get_results(...)`
 2) call a `hdc2010_trigger_conversion` to start the AMM, i.e. start measuring

every time there's a valid measurement by the chip, it'll call `user_cb`, where you call `hdc2010_get_results` - not results as no point of checking anything like which mode etc - so that you read out the new measurements.

File: drivers/include/hdc2010.h (hunk 5 above)
Comment (vigsterkr) [reply drivers/include/hdc2010.h:201]: indeed, it should be `uint16_t`

File: drivers/include/hdc2010.h (hunk 6 above)
Comment (vigsterkr) [reply drivers/include/hdc2010.h:186]: see about the reasoning why this is good to have it public

File: drivers/hdc2010/hdc2010.c (hunk 3 above)
Comment (vigsterkr) [reply drivers/hdc2010/hdc2010.c:123]: see below

File: drivers/include/hdc2010.h (hunk 5 above)
Comment (vigsterkr) [reply drivers/include/hdc2010.h:201]: although just a note: `phydat_t.val` is an `int16_t[3]`... hence there's gonna be a casting going on in saul if we change here to `uint16_t`. shouldn't be a problem since the value is between 0-10000, i.e. no negative value... just saying that imo it's not that nice that one is force to do an explicit cast there..

File: drivers/include/hdc2010.h (hunk 4 above)
Comment (benpicco) [reply drivers/include/hdc2010.h:139]: > every time there's a valid measurement by the chip, it'll call `user_cb`, where you call `hdc2010_get_results`

But you can't call `hdc2010_get_results` inside the callback since I2C operations are not allowed in interrupt context.

Instead of having the API user deal with that, why not simply do

```C
while (measuring) {
    hdc2010_read(dev, &temp, &hum);
    …
}
```

since `hdc2010_read()` is blocking you will get the periodic measurement you want.
You can also add a check to the function so it will only start the measurement once if periodic measurement is configured.
(There is no function to stop periodic measurement?)

**Comment by fjmolinas** [comment]
@benpicco @vigsterkr ping :)
